Parsing overhaul

bmacnaughton opened the pull request:
make parsing take less time than the httperf run

bmacnaughton committed:
rework parser

- add input data for testing

bmacnaughton committed:
tweak test to match timeit

bmacnaughton committed:
comment on regex setup

bmacnaughton committed:
README.txt - updates

cheempz commented on httperfpy/httperfpy.py:
```diff
@@ -117,55 +116,146 @@ def __params(self):
 
-    @classmethod
-    def parse(self, result_string):
+    def parse(self, result_string, options = {}):
```
Could you add some doc/comment about what `options` could be -- I know that right now it just allows seeing the transitions but even that might be useful for others when debugging.

bmacnaughton replied:
will do.

bmacnaughton replied:
done.

cheempz commented on httperfpy/httperfpy.py:
```diff
@@ -117,55 +116,146 @@ def __params(self):
+                if line.startswith(start):
+                    special_starts[start](line)
+                    found = True
+                if (found):
```
Is this if statement needed -- can the `break` just be added right after setting `found = True`?

bmacnaughton replied:
looks like i don't need it.

cheempz reviewed (commented):
Thanks for this overhaul @bmacnaughton!  I left some comments and then two main observations here:

* could you get the existing unit tests to work again (I think we can abandon python 2.7 and just support python 3).  And perhaps enhance with the httperf output samples added in this PR.
* there are a lot of unnecessary use of parens, you can run `pylint` on the source to get some pythonic advice.

bmacnaughton commented:
- working on unit tests
- i don't run pylint but run vscode's python linter and don't see unnecessary parens flagged. i'm not sure what you're referring to.

bmacnaughton committed:
README.txt - updates

bmacnaughton committed:
httperfpy.py - review comments changes

bmacnaughton committed:
unit - test updates

bmacnaughton committed:
httperf.py - fix missing state

bmacnaughton committed:
unit - update tests

- make work with python3
- fix for HttperfParser constructor
- test against expected results
- skip connection_times test
- add header test

bmacnaughton commented:
unit tests completed.
parens on if statements removed.

bmacnaughton committed:
use correct username

bmacnaughton committed:
httperfpy - comment typo

bmacnaughton committed:
httperfpy.py - remove async placeholder

bmacnaughton committed:
unit/parser_unit.py - skip comment fix

cheempz commented on httperfpy/httperfpy.py:
```diff
@@ -178,104 +268,197 @@ def __calulate_percentiles(self, pct, vct):
+
+            # Headers of interest
+            # select sampled headers only.
+            "header_xtrace": re.compile("^RH[0-9]+:X-Trace: ([0-9A-F]{59}1)", re.I)
```
This change will eliminate the need for loadgen.py to count the [sampled xtrace headers](https://github.com/librato/apm-benchmark-load-generators/blob/master/httperf/loadgen.py#L190) so will need corresponding tearing out of that code in loadgen.py.

bmacnaughton replied:
i hadn't asked you to review this yet, but: https://github.com/librato/apm-benchmark-load-generators/pull/4/files

bmacnaughton replied:
but it's not really necessary to change - loadgen will continue to work it will just count headers when it doesn't need to.

cheempz commented on unit/parser_unit.py:
```diff
@@ -8,13 +8,20 @@ class HttperfTestCase(unittest.TestCase):
+        presults = HttperfParser().parse(httperf_headers_results)
+        self.assertEqual(presults, httperf_headers_parsed)
+
+    @unittest.skip('not using jmervine/httperf')
```
OK, skipped tests get me curious so I tried it and got this:

```# python unit/parser_unit.py
.E..
======================================================================
ERROR: testParseVerboseResultCount (__main__.HttperfTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "unit/parser_unit.py", line 24, in testParseVerboseResultCount
    vpresults = HttperfParser().parse(httperf_verbose_results)
  File "/code/unit/../httperfpy/httperfpy.py", line 180, in parse
    special_starts[start](line)
  File "/code/unit/../httperfpy/httperfpy.py", line 139, in get_verbose_connection_time
    verbose_connection_times.append(match.group(1))
NameError: free variable 'match' referenced before assignment in enclosing scope
```

It look like coding error rather than test setup error with not having the special build (shouldn't really matter since it's testing with mocked data, not actual `httperf` output), so worth fixing even if we're not going to use the special `httperf` build right now.

bmacnaughton replied:
the main problem i had with it is that i cannot tell whether the Connection lifetime lines come 1) before the RH (header) lines, 2) after the header lines or 3) are interspersed with the header lines unless i install and run his modified httperf. i can make a very silly test work for the existing data but the code, as written now, will only work in 1 of those three cases.

bmacnaughton replied:
all that said, it's a coding error...

cheempz reviewed (approved):
LGTM pending the two comments I left.

bmacnaughton committed:
address review comments

bmacnaughton committed:
README.txt - updates

- recommend a container
- remove apt install python-unit note

bmacnaughton committed:
unit/unit_helper.py - change some values

- make some zero values non-zero
- verify that values are correct

bmacnaughton committed:
README.txt => README.md
